### ISSUE TITLE
How we can eliminate(remove) middleware in test mode (test api)?

### ISSUE THREAD

**Issue by pouryajafarzadeh** [tiangolo/fastapi#2495]
In fact, override the main middleware with another middleware (bypass the middleware) in a test mode.

**Comment by Mause** [comment]
Have you tried not adding the middleware in the first place? Then you don't have to remove it

**Comment by pouryajafarzadeh** [comment]
Yes. I could do that by removing middleware in the main app code, but I want to eliminate middleware in the test mode, when we are using:
`
from fastapi.testclient import TestClient
client =  TestClient(app)
`

**Comment by falkben** [comment]
yes, i think it's probably easier to never attach the middleware rather than try and remove it.

you could create the app in the test, and just leave off the middleware, just grabbing the router(s) (assuming you are using routers). that way your app is in main or wherever and the router is... wherever it is, but doesn't get any middleware attached to it.

or you could create a function to create the app which selectively mounts middleware  based on environment, so that you can still import and test the app w/o the middleware. something like the function proposed in [this comment](https://github.com/tiangolo/fastapi/issues/582#issuecomment-537015682).

curious though... wouldn't you want your tests to use the same middleware as the app does in production?

**Comment by pouryajafarzadeh** [comment]
Thank you. I find out your proposed solution.

**Comment by tiangolo** [comment]
Thanks for the help here @Mause  and @falkben  ! :clap: :bow: 

Thanks for reporting back and closing the issue @pouryajafarzadeh  :+1:

**Comment by mdryden** [comment]
I hope you don't mind a post on this old thread, but I found this issue while I was searching for solution to the same problem.  I didn't like the linked solution, but ended up working out my own.  Perhaps this will be useful to someone else:

```
test_client = TestClient(app)

app.user_middleware.clear()
app.middleware_stack = app.build_middleware_stack()
```

This seems to do what I (and probably the OP) was after.

**Comment by dmanthing** [comment]
`def remove_middleware(app: FastAPI, target: str) -> FastAPI:
    new_middlewares: list[Middleware] = []
    for middleware in app.user_middleware:
        if not middleware.cls.__name__ == target:
            new_middlewares.append(middleware)
    app.user_middleware = new_middlewares
    app.middleware_stack = app.build_middleware_stack()
    return app`
    
just pass the class name as `target` param :)

**Comment by Stblacq** [comment]
Heres an idea, if all you are trying to do is get past your middleware during test  
1.)  In your settings file    
```
import secrets
from pydantic import BaseSettings

class Settings(BaseSettings):
    ....
    APP_STATE_SECRET = secrets.token_hex(16)
```
2.   In your middle ware add   
 ```
 class AuthorizeRequestMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:  
....
        secret_key = request.headers.get('X_SECRET_KEY', False)
        if secret_key and secret_key == settings.APP_STATE_SECRET:
            return await call_next(request)
 ```  
3. Finally In your test   
```
from core.main import app
client = TestClient(app)


def test_auth_middleware():
    response = client.get("/", headers={'X_SECRET_KEY': settings.APP_STATE_SECRET})
    assert response.ok
```

**Comment by rajshahdev** [comment]
> ```
>     new_middlewares: list[Middleware] = []
```
from where should I import Middleware @dmanthing

**Comment by dmanthing** [comment]
> > ```
> >     new_middlewares: list[Middleware] = []
> > ```
> 
> ```
> from where should I import Middleware @dmanthing 
> ```

`from fastapi import FastAPI
from starlette.middleware import Middleware


def remove_middleware(app: FastAPI, target: str) -> FastAPI:
    new_middlewares: list[Middleware] = []
    for middleware in app.user_middleware:
        if not middleware.cls.__name__ == target:
            new_middlewares.append(middleware)
    app.user_middleware = new_middlewares
    app.middleware_stack = app.build_middleware_stack()
    return app`
